### ISSUE TITLE
Question: No need to use average gradients in a mini-batch?

### ISSUE THREAD

**Issue by AlexBYin** [jcjohnson/torch-rnn#129]
I might miss something here. 
The LSTM implementation here will take a batch of examples and during backward() gradients from each example are summed. The summed gradients are then sent to optim.adam(). 

I've seen other mini-batch SGD trainings where the summed gradients are divided by the batch size B. So the average gradient in the mini-batch is used in the optimization step.

This averaging only affects the magnitude of gradients, not the directions. It probably interacts with the learning rate because if you don't average then the gradients would be bigger in magnitude. 

Is there any particular reason torcn-rnn is not averaging the gradients in each mini-batch?

**Comment by jcjohnson** [comment]
The CrossEntropyCriterion used for training already averages the gradient over the minibatch:

https://github.com/torch/nn/blob/master/doc/criterion.md#crossentropycriterion

**Comment by AlexBYin** [comment]
Ah, I see. NNCriterion does the averaging. Thanks!
https://github.com/torch/nn/blob/master/lib/THNN/generic/ClassNLLCriterion.c
